File: product-microservice/api/hard_disk_drive_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type HardDiskDriveAPI struct {
	IHardDiskDriveService service.IHardDiskDriveService
}

func NewHardDiskDriveAPI(hardDiskDriveService service.IHardDiskDriveService) HardDiskDriveAPI {
	return HardDiskDriveAPI{IHardDiskDriveService: hardDiskDriveService}
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	hardDiskDrives := hardDiskDriveApi.IHardDiskDriveService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToHardDiskDriveDTOs(hardDiskDrives))
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetNumberOfRecords(c *gin.Context) {
	numberOfRecords := hardDiskDriveApi.IHardDiskDriveService.GetNumberOfRecords()
	c.JSON(http.StatusOK, numberOfRecords)
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetByID(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	hardDiskDrive, err := hardDiskDriveApi.IHardDiskDriveService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToHardDiskDriveDTO(hardDiskDrive))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (hardDiskDriveApi *HardDiskDriveAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	hardDiskDrives, err := hardDiskDriveApi.IHardDiskDriveService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToHardDiskDriveDTOs(hardDiskDrives))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := hardDiskDriveApi.IHardDiskDriveService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (hardDiskDriveApi *HardDiskDriveAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.HardDiskDriveFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	hardDiskDrives, err := hardDiskDriveApi.IHardDiskDriveService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToHardDiskDriveDTOs(hardDiskDrives))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetNumberOfRecordsFilter(c *gin.Context) {
	var filter filter.HardDiskDriveFilter
	err := c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	numberOfRecords := hardDiskDriveApi.IHardDiskDriveService.GetNumberOfRecordsFilter(filter)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetCapacities(c *gin.Context) {
	capacities := hardDiskDriveApi.IHardDiskDriveService.GetCapacities()
	c.JSON(http.StatusOK, capacities)
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetForms(c *gin.Context) {
	forms := hardDiskDriveApi.IHardDiskDriveService.GetForms()
	c.JSON(http.StatusOK, forms)
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetManufacturers(c *gin.Context) {
	manufacturers := hardDiskDriveApi.IHardDiskDriveService.GetManufacturers()
	c.JSON(http.StatusOK, manufacturers)
}

func (hardDiskDriveApi *HardDiskDriveAPI) GetDiskSpeeds(c *gin.Context) {
	diskSpeeds := hardDiskDriveApi.IHardDiskDriveService.GetDiskSpeeds()
	c.JSON(http.StatusOK, diskSpeeds)
}

func (hardDiskDriveApi *HardDiskDriveAPI) Create(c *gin.Context) {
	var hardDiskDriveDTO dto.HardDiskDriveDTO
	err := c.BindJSON(&hardDiskDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	hardDiskDrive := mapper.ToHardDiskDrive(hardDiskDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := hardDiskDriveApi.IHardDiskDriveService.Create(hardDiskDrive)

	if msg == "" {
		c.JSON(http.StatusOK, "Hard disk drive added successfully.")
	} else {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (hardDiskDriveApi *HardDiskDriveAPI) Update(c *gin.Context) {
	var hardDiskDriveDTO dto.HardDiskDriveDTO
	err := c.BindJSON(&hardDiskDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := hardDiskDriveApi.IHardDiskDriveService.Update(hardDiskDriveDTO)

	if msg == "" {
		c.JSON(http.StatusOK, "Hard disk drive updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (hardDiskDriveApi *HardDiskDriveAPI) Delete(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := hardDiskDriveApi.IHardDiskDriveService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "Hard disk drive deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/hard_disk_drive_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import hard_disk_drive_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_ALL
                        + "?page={}&pageSize={}".format(page, page_size))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_NUMBER_OF_RECORDS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_by_id(id):
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_BY_ID + "/{}".format(id))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.SEARCH_BY_NAME
                        + "?page={}&pageSize={}&name={}".format(page, page_size, name))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_search():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_NUMBER_OF_RECORDS_SEARCH)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    r = requests.post(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                         hard_disk_drive_api_routes.FILTER +
                         "?page={}&pageSize={}".format(page, page_size), json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_filter():
    data = request.json
    r = requests.post(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                         hard_disk_drive_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_manufacturers():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_MANUFACTURERS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_forms():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_FORMS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_capacities():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_CAPACITIES)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_disk_speeds():
    r = requests.get(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.GET_DISK_SPEEDS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    r = requests.post(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                         hard_disk_drive_api_routes.CREATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    r = requests.put(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                        hard_disk_drive_api_routes.UPDATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete(id):
    headers = request.headers
    r = requests.delete(hard_disk_drive_api_routes.BASE + hard_disk_drive_api_routes.API +
                           hard_disk_drive_api_routes.DELETE + "/{}".format(id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp
